internal/devices/assigned_devices.go

```diff
@@ -0,0 +1,163 @@
+// +build windows
+
+package devices
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+
+	winio "github.com/Microsoft/go-winio"
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/cmd"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+// HandleAssignedDevicesWindows does all of the work to setup the hosting UVM and retrieve
+// device information for adding assigned devices on a WCOW container definition.
+//
+// First, devices are assigned into the hosting UVM. Drivers are then added into the UVM and
+// installed on the matching devices. This ordering allows us to guarantee that driver
+// installation on a device in the UVM is completed before we attempt to create a container.
+//
+// Then we find the location paths of the target devices in the UVM and return the results
+// as WindowsDevices.
+func HandleAssignedDevicesWindows(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, specDevs []specs.WindowsDevice) (resultDevs []specs.WindowsDevice, closers []resources.ResourceCloser, err error) {
+	defer func() {
+		if err != nil {
+			// best effort clean up allocated resources on failure
+			for _, r := range closers {
+				if err := r.Release(ctx); err != nil {
+					log.G(ctx).WithError(err).Error("failed to release container resource")
+				}
+			}
+			closers = nil
+			resultDevs = nil
+		}
+	}()
+	vpciVMBusInstanceIDs := []string{}
+	for _, d := range specDevs {
+		if d.IDType == uvm.VPCIDeviceIDType || d.IDType == uvm.VPCIDeviceIDTypeLegacy {
+			vpci, err := vm.AssignDevice(ctx, d.ID)
+			if err != nil {
+				return nil, closers, errors.Wrapf(err, "failed to assign device %s of type %s to pod %s", d.ID, d.IDType, vm.ID())
+			}
+			closers = append(closers, vpci)
+			vmBusInstanceID := vm.GetAssignedDeviceVMBUSID(vpci.VMBusGUID)
+			log.G(ctx).WithField("vmbus id", vmBusInstanceID).Info("vmbus instance ID")
+
+			vpciVMBusInstanceIDs = append(vpciVMBusInstanceIDs, vmBusInstanceID)
+		} else {
+			return nil, closers, fmt.Errorf("device type %s for device %s is not supported in windows", d.IDType, d.ID)
+		}
+	}
+	if err := installWindowsDrivers(ctx, vm, annotations, closers); err != nil {
```

Comment:
It seems like this should instead be something like:

```
driverClosers, err := installWindowsDrivers(...)
if err ...
closers = append(closers, driverClosers)
deviceUtilClosers, err := setupDeviceUtilTool(...)
if err ...
closers = append(closers, deviceUtilClosers)
```

It's a lot more straightforward to receive a slice of closers as a return value, and append it to our master list, rather than passing a slice to be mutated into each function. Also, I don't think the current code will work properly, as we're passing the slice (ptr, len, cap) by value, rather than by pointer.

Reply:
What I don't like about this approach is that it becomes really unclear where device cleanup needs to happen.

Reply:
instead of resources being cleaned up in one place, now each function has to clean them up itself and because of the amount of refactoring already done, that gets messy and is duplicated a lot.

Reply:
Can you give an example of how resource cleanup would change? I'm not following.

Reply:
I would think that this change would require each called function to clean up the resources it has made. So for example in the format you mentioned: 

`func mountDrivers(ctx context.Context, vm *uvm.UtilityVM, hostPaths []string) ([]resources.ResourceCloser, []string, error) {
	resultClosers := []resources.ResourceCloser{}
	resultUVMPaths := []string{}
	for _, d := range hostPaths {
		share, uvmPath, err := addVSMBToUVM(ctx, vm, d)
		if err != nil {
			return nil, nil, err
		}
		resultClosers = append(resultClosers, share)
		resultUVMPaths = append(resultUVMPaths, uvmPath)
	}
	return resultClosers, resultUVMPaths, nil
}`

now it's unclear if this function should try to release the closers or just return whatever closers it made. If it doesn't clean up it's closers, now it's the caller's responsibility to make sure that they're cleaned up even if 
`mountDrivers` returns an error. Either way, the caller would now need to do the below since there could be errors from other functions: 

`func InstallWindowsDrivers(ctx context.Context, vm *uvm.UtilityVM, drivers []string) (closers []resources.ResourceCloser, err error) {
	defer func() {
		if err != nil {
			for _, r := range closers {
				if releaseErr := r.Release(ctx); releaseErr != nil {
					log.G(ctx).WithError(releaseErr).Error("failed to release container resource")
				}
			}
		}
	}()
	closers, driverUVMPaths, err := mountDrivers(ctx, vm, drivers)
	if err != nil {
		return closers, err
	}
	return closers, execPnPInstallAllDrivers(ctx, vm, driverUVMPaths)
}`

But then additionally, the caller of InstallWindowsDrivers, `handleAssignedDevicesWindows` would need to do something similar since it calls into other functions that could fail as well. So now we've had to clean up in three different spots for the same error.

Does that make sense?

Reply:
And if we just have each function propagate the closers up whether or not the function failed (ie the called function doesn't do resource cleanup), now we need to do the below to make sure that all possible closers created are released later: 

```
                share, uvmPath, err := addVSMBToUVM(ctx, vm, d)
		resultClosers = append(resultClosers, share) // THIS MUST GO BEFORE RETURNING
		if err != nil {
			return resultClosers, resultUVMPaths, err
		}
```

Reply:
I made the change. Please take a look.